File: design-add-and-search-words-data-structure/WordDictionary.py
# https://leetcode.com/problems/design-add-and-search-words-data-structure/

class TrieNode:

    def __init__(self):
        self.children = [None] * 26
        self.is_end_of_word = False


class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    # T: O(n), S: O(n)
    def addWord(self, word: str) -> None:
        curr = self.root
        for ch in word:
            idx = ord(ch) - ord('a')
            if curr.children[idx] is None:
                curr.children[idx] = TrieNode()
            curr = curr.children[idx]
        curr.is_end_of_word = True

    # T (best no wildcards): O(n)
    # T (worst with wildcards): O(26^n)
    # S: O(n)
    def search(self, word: str) -> bool:
        return self._search_helper(word, 0, self.root)

    # T (best no wildcards): O(n)
    # T (worst with wildcards): O(26^n)
    # S: O(n)
    def _search_helper(self, word: str, idx: int, node: TrieNode) -> bool:
        if idx == len(word):
            return node.is_end_of_word

        ch = word[idx]

        if ch == '.':
            for child in node.children:
                if child is not None and self._search_helper(word, idx + 1, child):
                    return True
            return False

        char_idx = ord(ch) - ord('a')

        if node.children[char_idx] is None:
            return False

        return self._search_helper(word, idx + 1, node.children[char_idx])


def main() -> None:
    wd = WordDictionary()
    wd.addWord("bad")
    wd.addWord("dad")
    wd.addWord("mad")
    print(wd.search("pad"))  # false
    print(wd.search("bad"))  # true
    print(wd.search(".ad"))  # true
    print(wd.search("b.."))  # true


if __name__ == "__main__":
    main()

File: design-add-and-search-words-data-structure/WordDictionary.java
// https://leetcode.com/problems/design-add-and-search-words-data-structure/

class WordDictionary {

    class TrieNode {
        TrieNode[] children;
        boolean isEndOfWord;

        TrieNode() {
            children = new TrieNode[26];
            isEndOfWord = false;
        }
    }

    private TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }

    // T: O(n), S: O(n)
    public void addWord(String word) {
        TrieNode curr = root;
        for (char ch : word.toCharArray()) {
            int idx = ch - 'a';
            if (curr.children[idx] == null) {
                curr.children[idx] = new TrieNode();
            }
            curr = curr.children[idx];
        }
        curr.isEndOfWord = true;
    }

    // T (best no wildcards): O(n)
    // T (worst with wildcards): O(26^n)
    // S: O(n)
    public boolean search(String word) {
        return searchHelper(word, 0, root);
    }

    // Recursive helper method for search with wildcard support
    // T (best no wildcards): O(n)
    // T (worst with wildcards): O(26^n)
    // S: O(n)
    private boolean searchHelper(String word, int idx, TrieNode node) {
        // Base case: reached the end of the word
        if (idx == word.length()) {
            return node.isEndOfWord;
        }

        char ch = word.charAt(idx);

        // Handle wildcard '.' character
        if (ch == '.') {
            // Try all possible children
            for (TrieNode child : node.children) {
                if (child != null && searchHelper(word, idx + 1, child)) {
                    return true;
                }
            }
            return false;
        }

        // Regular character search
        int charIdx = ch - 'a';

        // If character node doesn't exist, return false
        if (node.children[charIdx] == null) {
            return false;
        }

        // Recursive search for the next character
        return searchHelper(word, idx + 1, node.children[charIdx]);
    }

    public static void main(String[] args) {
        var wd = new WordDictionary();
        wd.addWord("bad");
        wd.addWord("dad");
        wd.addWord("mad");
        System.out.println(wd.search("pad")); // false
        System.out.println(wd.search("bad")); // true
        System.out.println(wd.search(".ad")); // true
        System.out.println(wd.search("b..")); // true
    }

}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
